### ISSUE TITLE
VS Code crashes when running a cell that produces a lot of output

### ISSUE THREAD

**Issue by ydmykr** [microsoft/vscode-jupyter#11031]
I am trying to print model summary in tensorflow and I think the model is large and it's crashing my notebook. The model is ResNet101.

![image](https://user-images.githubusercontent.com/100815370/182506283-b4d560f0-7f6a-40de-9ef8-27ae7c245bf2.png)

The whole computer comes to a halt, memory usage goes up to 99% and VS Code crashes. I have 16 GB of ram, so I didn't think printing something large would actually eat all my ram. Also, because the kernel crashes, all the variables are lost like history = model.fit() which I need to fine-tune the model afterwards. Moreover, I need to print base_model summary in order to choose from which layer to fine-tune from.

Is there a way to print the summary in another way and save the entire notebook with the variables, so I can continue working? I have checkpoints for model weights, but I need to keep track of past epochs through history to resume training afterwards.

I will try using`.py` files, but I want to know if there is a way to solve this problem for jupyter.

**Comment by DonJayamanne** [comment]
Thanks for filing this issue and I'm sorry you are running into this.
* Is it VS Code or is it is the kernel that crashes?
* If you were to run this outside VS Code in classic jupyter via a terminal , do you experience the same problem?
* Can you run a cell at a time to check when exactly vscode crashes, that would help narrow down the issue

**Comment by DonJayamanne** [comment]
I'm assuming you are referring to the cell with the code `base_model.summary`
Is it possible to try the following:
* Not print any output at all
* Run the first cell, then clear all output using the command or icon in the toolbar
* Run the next cell and clear all of the output.
* Similarly continue till you get to the problematic cell, again, comment out the code where you are printing the model summary.

What I'd like to figure out is:
* Whether the memory usaage/crash is realted to the output or execution
* If we ensure the output is always cleared, and things work as expected (i.e. it does not crash), then we've narrowed down the issue to outputs causing increased memory usage and crashes vscode.

Finally
* Also, please could you share the output from the `Jupyter` output panel, use the command `Jupyter: Show Output`
Run the first cell in the notebook and copy the logs and send the logs (please ensure you send the whole log). Shows informatino about versions, python used, and the like.
* Are you running the code against a local or remote kernel?
* > memory usage goes up to 99% and VS Code crashes. I
Please could you take a screenshot of this and upload that here, would like to check what vscode process is chewing up the memory. thanks

**Comment by mariosconsta** [comment]
I am the OP who posted this question on stack overflow and OP here created the issue for me, so I will answer your questions.
So after making a new notebook to test again, the same thing happens:
![296595032_738041494121566_6923916001967281265_n](https://user-images.githubusercontent.com/22726730/182798388-e741cec0-7be9-493b-881a-66bb052ff1bb.jpg)
Sorry for the phone screenshot, computer froze. Also, please note that because the computer is getting so slow, it can't show me the exact usage number, but I guess it's gigabytes since its saying 99%
![Screenshot 2022-08-04 111613](https://user-images.githubusercontent.com/22726730/182798629-eaa2b412-08e0-4903-8dfc-28b6e173430c.png)

-  Clearing outputs is not changing the outcome

> I have a suspicion the output from the second cell is very large & that could be chewing up resources.
Again, please ensure you clera all of the output after running each cell.

- I have put verbose = 2 for model.fit so that output is not large. But eitherway, the same thing happens on another notebook where I do not train any model and the only output is base_model.summary()

> Similarly continue till you get to the problematic cell, again, comment out the code where you are printing the model summary 

Yep the problematic cell is the one where I print the summary(). Same thing happens on a brand new and small notebook that I just import ResNet101 and try to print it's summary.

> What I'd like to figure out is:

> Whether the memory usaage/crash is realted to the output or execution

I think it's execution and not output. Because even if 15-20secs has passed since I pressed run on that cell, after it finishes it shows 0.5s and when I run that cell again, it is taking less time which I believe is because the variable is already saved and it doesn't need to compute it again.
![image](https://user-images.githubusercontent.com/22726730/182800738-0abff076-580f-4501-b813-38eb3b425d8b.png)

> If we ensure the output is always cleared, and things work as expected (i.e. it does not crash), then we've narrowed down the issue to outputs causing increased memory usage and crashes vscode.

Unfortunately, that's not the issue

> Are you running the code against a local or remote kernel?

Local Kernel

> Also, please could you share the output from the Jupyter output panel, use the command Jupyter: Show Output
Run the first cell in the notebook and copy the logs and send the logs (please ensure you send the whole log). Shows informatinon about versions, python used, and the like.

[logs.txt](https://github.com/microsoft/vscode-jupyter/files/9257985/logs.txt)

**Comment by DonJayamanne** [comment]
Thanks for all of the information, please could you:
* Provide a very small sample code to replicate this at our end.
You mentioned a small notebook, please provide that as a sample or something even smaller.
* Please provide all of the necessary python packages that need to be installed, if you are using conda, then even better as you could just  provide the conda yaml file (do let us know the version of the python environment as well).
* If there are ay other files that we need, please could you provide those as well (feel free to email them to me, if you think they are they shoudl not be shared online, my email is on my github profile)

**Comment by mariosconsta** [comment]
> Thanks for all of the information, please could you:
> 
> * Provide a very small sample code to replicate this at our end.
>   You mentioned a small notebook, please provide that as a sample or something even smaller.
> * Please provide all of the necessary python packages that need to be installed, if you are using conda, then even better as you could just  provide the conda yaml file (do let us know the version of the python environment as well).
> * If there are ay other files that we need, please could you provide those as well (feel free to email them to me, if you think they are they shoudl not be shared online, my email is on my github profile)

The zip folder should contain everything you need. The small demo notebook at around 100KB and a notepad with all my packages, you can run pip install with that notepad and it will install everything for you.

[demo.zip](https://github.com/microsoft/vscode-jupyter/files/9258125/Desktop.zip)

Also I am using venv, python version is 3.10.2, tensorflow version is the latest. If you can reproduce it, I can take a video of my phone and show you step by step what is happening.

**Comment by DonJayamanne** [comment]
@mariosconsta 
I've tried running this at my end, both on a Mac and Windows machine and have no problems.

Can you try the following:
* Step 1
    * Run the notebook in Jupyter notebooks, save the notebook and upload that saved notebook here.
    > notebook where I do not train any model and the only output is base_model.summary()
    * Run the above notebook in Jupyter notebooks as well , save the notebook and upload that saved notebook here.
* Step 2
    * Exit vscode completely
    * Open the above two notebooks and see if you have similar perf issues in vscode.
    * Check the memory usage
    * Note: At this point, you are NOT running anything in VS Code, just opening the notebooks

Basically I'd like to get the two notebooks.

**Comment by mariosconsta** [comment]
I will test this out ASAP and let you know.

**Comment by mariosconsta** [comment]
@DonJayamanne 

So with VS Code **closed**. I opened jupyter notebook and run the cell that outputs model summary. The cell ran instantly without trouble. I opened VS Code to try it again and the same thing happened.

[notebooks.zip](https://github.com/microsoft/vscode-jupyter/files/9375366/notebooks.zip)

I will take a video with my phone as its happening and upload it here as well.

**Comment by mariosconsta** [comment]
@DonJayamanne 
I have uploaded 2 clips of the problem. The second clip is from my phone because I couldn't record normally. The first clip is showing jupyter running like a charm and the second one is VS Code consuming 100% of my RAM and then crashing.

There's also 2 screenshots on how VS Code looks after it recovers along with the error that sometimes pops up. The error doesn't pop up always.

https://drive.google.com/file/d/1GzOu8ZdX9CTZeN7guRcpZpYaWco5B445/view?usp=sharing

**Comment by DonJayamanne** [comment]
Will get a custom build of he extension shortly, so you can test with that
Once again, thanks for your patience and sorry for the inconvenience caused.

**Comment by DonJayamanne** [comment]
Please could you:
* Install VS Code Insiders
* Install the attached vsix using the command `Extnsions: Install from VSIX`
* Reload VS Code insiders
* Try running the notebook with this version
* Note: You might not see any outputs in any cell, however you should still see the green `tick` to indicate whether a cell ran successfully or not, along with the execution count

Please let me know how this goes.
[ms-toolsai-jupyter-insiders.vsix.zip](https://github.com/microsoft/vscode-jupyter/files/9377909/ms-toolsai-jupyter-insiders.vsix.zip)

**Comment by mariosconsta** [comment]
I will test it out first thing in the morning!

**Comment by mariosconsta** [comment]
@DonJayamanne Runs instantly as shown in the screenshot below. Like you said there's no output, but it does run like a charm and uses zero ram. How do should I continue? Is there a way to get output from this version? Or should I continue using the version I currently am on.

![image](https://user-images.githubusercontent.com/22726730/185584335-7d3511cf-748a-48b9-8a2f-c3d533232619.png)

**Comment by mariosconsta** [comment]
Should I re-install VS Code? Is there a way to save my configuration (extensions and local settings)? I am in the middle of my thesis, and I am running low on time, so it would be ideal if I could install it again and have all my extensions and settings as they are.

**Comment by mariosconsta** [comment]
Another issue is this:

![image](https://user-images.githubusercontent.com/22726730/185592636-4a282659-bcf7-40b1-b365-7a26fc1855dc.png)

Sometimes when it finishes without crashing, the output looks like that.

**Comment by DonJayamanne** [comment]
@mariosconsta Please could you try installing this VSIX as well and letting me know whether this still crashes.
Thanks again for your help and patience
[ms-toolsai-jupyter-insiders.vsix (1).zip](https://github.com/microsoft/vscode-jupyter/files/9385162/ms-toolsai-jupyter-insiders.vsix.1.zip)

**Comment by DonJayamanne** [comment]
> I am willing to do an AnyDesk session with you if you want to see it "live" if the recordings I sent you were not enough.

Thanks, so far what we have going is working, i'm still trying to gather some more information before I start working on this issue.

**Comment by mariosconsta** [comment]
> Please could you try installing this VSIX as well and letting me know whether this still crashes.

Just like the last VSIX, works like a charm. The cell runs instantly without issues.

![image](https://user-images.githubusercontent.com/22726730/185705224-f6ae2753-fdc8-40be-b006-9b0156d2701a.png)

**Comment by mariosconsta** [comment]
>Please could you try installing this VSIX as well and letting me know whether this still crashes.

Also like before, no output. Could this be the issue? The output? I did some testing and I was able to output large text without issue using a for loop.

You can check the recording here:
[demo2.zip](https://github.com/microsoft/vscode-jupyter/files/9385382/demo2.zip)

**Comment by DonJayamanne** [comment]
Internal note: Blocked on upstream issue https://github.com/microsoft/vscode/issues/146768

> Also like before, no output. Could this be the issue? The output? I did some testing and I was able to output large text without issue using a for loop.

@mariosconsta I'm sorry, but I dont follow you.
From what I understand with the VSIX i provided, it doesn't crash anymore, however hot output is displayed is that right.


> Could this be the issue? The output? I 

Currently the extension host seems to crash due to the size of the output generated.

**Comment by mariosconsta** [comment]
> I'm sorry, but I dont follow you.

Both VSIX you provided do not crash but do not show any output either.

> Currently the extension host seems to crash due to the size of the output generated.

I see, well let me know if you need anything from me and if there's any way to work around this problem.

**Comment by mariosconsta** [comment]
> Currently the extension host seems to crash due to the size of the output generated.

@DonJayamanne by the way, if you saw the recording in the demo2.zip, I printed only the layers of the model using a for loop, which is a very lost list of text, and it runs instantly.

**Comment by DonJayamanne** [comment]
@mariosconsta
Please could you upload the `dataset` folder as well. That wasn't in the zip file you sent earlier.

**Comment by DonJayamanne** [comment]
@mariosconsta 
Thanks a lot for your help and patience, I'm trying to resolve this issue in VS Code, and have made some progres (which is why the issue was accidentally closed via a PR).
However I still believe the fix will bot be sufficient.
* Attached is a VSIX, 
* Please install the latest VS Code Insider version and this VSIX
* Open your above problematic notebook that crashes VS Code
* Run just that ONE notebook (run all cells or what ever steps you use to crash VS Code)
* A couple of text file will be created next to the notebook of the form `<notebook file name><number>.log`
* Please upload the log files and the ipynb file you used

These log files contain all of the output generated and added to the notebook, hence should  not contain any personal information, it would only contain the same information taht wouuld be stored in the in the ipynb files as output. 
If you notebook file is at `C:\temp\one.ipynb`, then the logs files would be `C:\temp\one.ipynb1.log`, `C:\temp\one.ipynb2.log` & so on. One file for each cell.

If you are going to edit the log files, please let me know so that I can account for those.

I'd really appreciate it if you could provide these logs, that would help me replicate this at my end without any of the additonal packages.
[ms-toolsai-jupyter-insiders.vsix.zip](https://github.com/microsoft/vscode-jupyter/files/9552868/ms-toolsai-jupyter-insiders.vsix.zip)


Thanks again

**Comment by DonJayamanne** [comment]
> your GitHub profile to send you the Dataset tha
<first name>.<last name><at symbol><company I work for><dot><com>
First name = Don
Last Name = Jayamanne

Thanks a lot, will check the logs

**Comment by mariosconsta** [comment]
@DonJayamanne Sending you the dataset through WeTransfer right now.

**Comment by DonJayamanne** [comment]
@mariosconsta I'm sorry we weren't able to get this fixed in the current iteration, we've been working on this issue, however it ended up taking a bit longer than expected (due to changes in the API and fine turning the code). Our plan is to get this addressed in the next release.

**Comment by mariosconsta** [comment]
@DonJayamanne No worries, let me know if you need anything else.

**Comment by codeananda** [comment]
Just read through this whole thread. I've experienced the same issue when running a long bit of code outputting plots overnight.

It crashed last night (indeed, it has crashed every time I've tried to run it). 

Now two notebooks seem to be corrupted. One I was running last night (notebook 1) and the other (notebook 2) which also contains a lot of plots. I was working in notebook 2 yesterday but it was closed during the overnight crash. White rendering over all markdown cells and outputs cannot be seen. If you create a new code cell and execute, it doesn't produce any output.

Is there any way to recover the notebooks? I've tried opening them on VS Code Insiders but they are also broken there. However, when opening in native jupyter, the output of notebook 1 for the cell that caused the crash is gone. For notebook 2, all the plots are still there (which is good news!)

<img width="626" alt="image" src="https://user-images.githubusercontent.com/51246969/200828474-f51fddf5-f659-4783-aea9-4889efbe9943.png">

**Comment by mariosconsta** [comment]
Yeah I can verify as well that it does not work. I run my notebook again and the whole computer froze for 3 solid minutes and then:
![image](https://user-images.githubusercontent.com/22726730/201540296-9931dc00-3344-4dee-a5e6-e7175c06acbf.png)

**Comment by DonJayamanne** [comment]
hi everyone which, but this should be fixed in the next release (we were unable to take last month).

you can try installing the latest pre-release version of the jupyter extension and that should work.

**Comment by dokutoshi** [comment]
The issue remains in the latest VScode.  I am running a long Resnet mode to a remote server using VScode.  Vscode crashed twice.
VScode Jupyter is a white wall as described earlier. Just to remind you, I am not running the code locally but remotely.
![Image 2-27-23 at 10 31 PM](https://user-images.githubusercontent.com/10182821/221746580-af584879-eab8-4575-b1ab-f300eabf68aa.jpg)

**Comment by codeananda** [comment]
Update: I haven't had the white walls for a while. I've switched to putting all my files on the WSL filesystem and have been using a docker container as my development environment. I've also moved away from using notebooks as much and focused on coding in py files and using notebooks more like scratchpads to check bits of code work or to inspect plots. 

Switching all files to WSL had a massive boost on performance in general. 

Not sure if that's helpful but it's what has been working for me for time being.

**Comment by DonJayamanne** [comment]
@mariosconsta @codeananda @dokutoshi 
I'm sorry you are still running into this issue.
* Assume you save the notebook and open it again in VS Code, does VS Code then crash again?
* Assume you run this notebook in regular Jupyter Notebook/Lab, and then open the notebook in VS Code, does it crash again

Basically trying to figure out whether its caused by
* The regular streaming of data 
* Or because the output in a cell is too large

Please could you help us repro this issue
* Could you perhaps create a docker file with the necessary dependencies with instruction so we can test this at our end
* Please remember to include the same code (notebook) with instructions for installing all dependencies

> just tried to do a big calculation in a notebook and the White Walls of Doom™ came back. It seems like VS Code + Notebooks + Big Calculations = Bad Idea.

@codeananda 
Is this in docker or WSL. Please could you share some sample code

Basically I'm looking for some sample code that would generate such large content to cause this issue at our end

**Comment by dokutoshi** [comment]
Will respond today. 

For now, enumerations i.e. printing accuracy definitely affect memory.  
Here is a screenshot of a G4dn AWS EC2 running an inference on 75,000 images.  

The server is utilizing 53GB of RAM mostly because of VSCode server's nodeJS.  
The MacBook Pro M1, 16GB RAM uses about 2GB to 10GB to manage the Remote SSH.  

![image](https://user-images.githubusercontent.com/10182821/222549615-c6d03afb-3ffd-46f8-9282-27654d6007ba.png)

**Comment by DonJayamanne** [comment]
Thanks, hopefully you can share something that can be run locally. Mac M1 is also fine

**Comment by tyePhDCandy** [comment]
same problem here. 

vscode occupies like 45% of my cpu computing power. my cpu is amd 5700g and i have 64GB of memory. 

I think the problem lies in the way of vscode-jupyter handles the output info when they are thousands of them. 

i dont think the occassion is very easy to replicate, because it only happens on one of my notebook file. 

I will report more, if i have the time.

**Comment by livan3li** [comment]
hi, has this problem been solved? @DonJayamanne 

I have same problem right now. VS Code consumes all available memory and OS crashes if i dont close the VS CODE  When i try to open a Jupyter Notebook  which have 618 images in.

But weird thing is that when i try to open same notebook in Google Colab everything is perfect. No memory consumption and even more i can view all images very fast by sliding scroll bar very very fast and as a result no freeze, no stuck, no problem. 

I think the problem might be all about loading all data into memory(or memory management), not from storage device(HDD/SSD). Question is how Google Colab can achive such a good performance.

Thanks.

**Comment by DonJayamanne** [comment]
@livan3li  
Unfortunately this hasn't been fixed yet
please could you share this notebook
I'd like to try this at my end

Also, pleat could your disable all extensions, including Jupyter and see if you run into this issue
If you do then the problem is different, as this issue focuses on crashes doe to running of notebooks
However I do want to address your issue too, but would like to deal with that separately, for now let's chat here

**Comment by livan3li** [comment]
Ok. I've disabled all extensions related with Jupyter after this it's been opened as plain text file. 

Here is the [notebook link](https://drive.google.com/file/d/1-QJsiHnzgbhvD6KudTEzgUuFfHaEmMOL) in drive which i have used in the video. 

> @livan3li Unfortunately this hasn't been fixed yet please could you share this notebook I'd like to try this at my end
> 
> Also, please could your disable all extensions, including Jupyter and see if you run into this issue If you do then the problem is different, as this issue focuses on crashes doe to running of notebooks However I do want to address your issue too, but would like to deal with that separately, for now let's chat here

**Comment by DonJayamanne** [comment]
>  I've disabled all extensions related with Jupyter after this it's been opened as plain text file.

What i meant was, disable all extensions (including Jupyter), but still open as a notebook.
You can still open the ipynb file as a notebook even with teh Jupyter extension disabled,
Please try to re-open the file as a notebook (instead of opening it as a plain text file)
Opening and rendering notebooks is a feature that is a core functionality in vscode

**Comment by DonJayamanne** [comment]
@rebornix aren't you already looking into some perf improvements for opening notebooks with large number of outputs?

**Comment by livan3li** [comment]
> > I've disabled all extensions related with Jupyter after this it's been opened as plain text file.
> 
> What i meant was, disable all extensions (including Jupyter), but still open as a notebook. You can still open the ipynb file as a notebook even with teh Jupyter extension disabled, Please try to re-open the file as a notebook (instead of opening it as a plain text file) Opening and rendering notebooks is a feature that is a core functionality in vscode

Sorry for that. I've done exactly what you said. I disabled all extensions (even Pyhton) and tried to open the notebook but same thing happened again. VS Code consumes all available memory.

**Comment by tyePhDCandy** [comment]
> Yeah I can verify as well that it does not work. I run my notebook again and the whole computer froze for 3 solid minutes and then: ![image](https://user-images.githubusercontent.com/22726730/201540296-9931dc00-3344-4dee-a5e6-e7175c06acbf.png)

this happens to me too.

**Comment by livan3li** [comment]
> > Yeah I can verify as well that it does not work. I run my notebook again and the whole computer froze for 3 solid minutes and then: ![image](https://user-images.githubusercontent.com/22726730/201540296-9931dc00-3344-4dee-a5e6-e7175c06acbf.png)
> 
> this happens to me too.

Since i disabled all extensions, Python doesn't do that for me. But now under VS Code, again VS Code consumes all memory.

**Comment by livan3li** [comment]
> > > > I've disabled all extensions related with Jupyter after this it's been opened as plain text file.
> > > 
> > > 
> > > What i meant was, disable all extensions (including Jupyter), but still open as a notebook. You can still open the ipynb file as a notebook even with teh Jupyter extension disabled, Please try to re-open the file as a notebook (instead of opening it as a plain text file) Opening and rendering notebooks is a feature that is a core functionality in vscode
> > 
> > 
> > Sorry for that. I've done exactly what you said. I disabled all extensions (even Python) and tried to open the notebook but same thing happened again. VS Code consumes all available memory.
> 
> you can check the subprocess under main process 'vscode' . see which subprocess consumes all that memory. I found out, on my end, the python subprocess under 'vscode' consumes loads of memory.

Since i disabled all extensions, Python doesn't do that for me. But now under VS Code, again VS Code consumes all memory.

**Comment by livan3li** [comment]
> > I've disabled all extensions related with Jupyter after this it's been opened as plain text file.
> 
> What i meant was, disable all extensions (including Jupyter), but still open as a notebook. You can still open the ipynb file as a notebook even with teh Jupyter extension disabled, Please try to re-open the file as a notebook (instead of opening it as a plain text file) Opening and rendering notebooks is a feature that is a core functionality in vscode

Have you reported bug where it needs to be or i need to do that?

**Comment by DonJayamanne** [comment]
@livan3li @tyePhDCandy 
Thanks for continuing to provide feedback, however it gets a little confusing when we try different things
Please could you confirm the following:
* Disable all extensions
* Close VS Code
* Re-open VS Code
* Open the same large notebook
* VS Code hangs again

This will confirm that the freeze is due to loading of the notebook
The memory usage of the python process could be due to the large training job or the like.
Optionally you could try the same in Jupyter Notebook Application outside VS Code and see if there are similar high memory usages by that application, which would indicate the fact that it is expected (due to the code being executed)

**Comment by Megil-Zesty** [comment]
Also experiencing this issue - super frustrating.

Context: notebook essentially freezes when I print too much to a console or when trying to display a large pandas dataframe (by simply typing "df" in a cell and running it"). I have 104gb of ram and the dataframes are only consuming about 7gb.

**Comment by DonJayamanne** [comment]
@Megil-Zesty 
* Please could you share some sample code that can be used to replicate this issue.

**Comment by rebornix** [comment]
This issue should now be fixed in latest Insiders and will be available in next week's Stable release. We will revisit this issue once you all get the latest fixes and reopen if necessary.

**Comment by jrieken** [comment]
Tested successfully with a dummy controller that produces 10000000 lines textual output. That might have done but having author verification or crisp verification steps would help

**Comment by joserojasb** [comment]
Hi everyone! I've run into the white walls of doom just after a recent vscode update, on my Mac M1 Pro. Due to compliance reasons i'm unable to provide code examples, but some of the issues others have reported seem to keep popping up: 

- It happens when running exhaustive calculations/displaying long outputs
- Other notebooks that previously ran fine seem now to be "corrupted", or at least this is a persistent issue even when rebooting
- I've purged vscode and reinstalled everything, to no avail

I hope this information is at least a little bit useful!

**Comment by r3m0t** [comment]
I have a repro here https://github.com/microsoft/vscode/issues/182636 @jrieken

**Comment by johnomage** [comment]
I currently having this issue with version 1.81.0 😭😞🤬
